shotover-proxy/src/transforms/protect/encrypt.rs

```diff
@@ -0,0 +1,66 @@
+use crate::message::MessageValue;
+use crate::transforms::protect::key_management::KeyManager;
+use anyhow::{anyhow, bail, Result};
+use serde::{Deserialize, Serialize};
+use sodiumoxide::crypto::secretbox;
+use sodiumoxide::crypto::secretbox::Nonce;
+use sqlparser::ast::Value as SQLValue;
+
+// A protected value meets the following properties:
+// https://doc.libsodium.org/secret-key_cryptography/secretbox
+// This all relies on crypto_secretbox_easy which takes care of
+// all padding, copying and timing issues associated with crypto
+#[derive(Serialize, Deserialize)]
+struct Protected {
+    cipher: Vec<u8>,
+    nonce: Nonce,
+    enc_dek: Vec<u8>,
+    kek_id: String,
+}
+
+pub async fn encrypt(
+    value: &SQLValue,
+    key_management: &KeyManager,
+    key_id: &str,
+) -> Result<SQLValue> {
+    let value = MessageValue::from(&*value);
+
+    let sym_key = key_management.cached_get_key(key_id, None, None).await?;
+
+    let ser = bincode::serialize(&value)?;
+    let nonce = secretbox::gen_nonce();
+    let cipher = secretbox::seal(&ser, &nonce, &sym_key.plaintext);
+
+    let protected = Protected {
+        cipher,
+        nonce,
+        enc_dek: sym_key.ciphertext_blob.to_vec(),
+        kek_id: sym_key.key_id,
+    };
+
+    Ok(SQLValue::SingleQuotedString(serde_json::to_string(
```

Comment:
We used to store the Protected value in a blob / bytes type. Admittedly the prior state of protect was broken on the serialization side of things. Whats the rationale for storing as a string vs bytes?

Reply:
I dont think we ever stored as a blob as there was no way to do that with the old sqlparser implementation.
That will become possible once we move to cql3-parser though.

Reply:
oh, wait, you are correct actually!

What we did was use MessageValue::Bytes in protect.
However shotover was internally converting that into a SQLValue::SingleQuotedString

What I will do is revert the change here to use a MessageValue::Bytes and then let the cql3-parser PR resolve the bytes/string issues.

Reply:
actually, actually.
This specific line is fine.

This is equivalent to the old logic where MessageValue::Bytes was internally converted to a SQLValue::SingleQuotedString.
However now we are just directly using the SingleQuotedString.

However there is a line in `fn decrypt` that was expecting a MessageValue::Varchar instead of a MessageValue::Bytes, so I changed that to MessageValue::Bytes.
decrypt is broken anyway, but at least this is closer to something that works.

As before though, the cql3-parser PR actually fixes this brokenness and has a full integration test proving that encryption/decryption round trips correctly.

Reply:
ok cool